osbrain/helper.py

```diff
@@ -192,3 +192,36 @@ def wait_agent_attr(agent, name='received', length=None, data=None, value=None,
             break
         time.sleep(0.01)
     return False
+
+
+def synchronize_sync_pub(server, server_alias, client, client_alias):
+    '''
+    Create a SYNC_PUB/SYNC_SUB channel and connect both agents.
+
+    Make sure they have stablished the PUB/SUB communication within the
+    SYNC_PUB/SYNC_SUB channel before returning. This will guarantee that
+    no PUB messages are lost.
+    '''
+    def assert_receive(agent, message, topic=None):
+        try:
+            agent.get_attr('_tmp_attr')
+            agent.set_attr(_tmp_attr=True)
+        except AttributeError:  # Attribute already deleted
+            pass
+
+    uuid = client.get_uuid_used_as_alias_for_sub_in_sync_pub(client_alias)
+
+    # Set a temporary custom handler
+    client.set_attr(_tmp_attr=False)
+    original_handler = client.get_handler(uuid)
+    client.subscribe(uuid, assert_receive)
+
+    # Send messages through the PUB socket until the client receives them
+    server.each(0.1, 'send', server_alias, 'Synchronize', alias='_tmp_timer')
+    assert wait_agent_attr(client, name='_tmp_attr', value=True, timeout=5)
+    server.stop_timer('_tmp_timer')
+
+    # Restore the original handler, now that the connection is guaranteed
+    client.subscribe(uuid, original_handler)
+
+    client.del_attr('_tmp_attr')
```

Comment:
Hmmmmm.

Yeah, one possibility would be to take that away from there and keep it just for the tests.

Another possibility would be to make all SYNC_PUB sockets implement heart-beating. This is something we should do at some point anyway, hopefully Soon™.

Heart-beating would mean that a SYNC_PUB socket would publish a "heart-beat" (like an "I'm alive" message) periodically, to all subscribed agents. Clients can use this to know if they should reconnect (maybe the server went down).

What we could do for now is to implement just a part of the heart-beating: allow clients to request a heart-beat at any time (i.e.: "are you alive?"). This would be a request from the SYNC_SUB, although it would be treated as a special case and all SYNC_PUB channels should handle it the same way, without taking into account the user-defined handler for other requests. This means adding a "special case" in the protocol, but I don't think it is bad, as requests do not need to be very fast and are infrequent.

We could then use those heart-beat requests from the tests to ensure synchronization.

If you feel like that is too much for now, just take it away from there and open a new issue so we try to improve it for version 1.0.0.

![](http://cdn.toonvectors.com/images/2/3529/toonvectors-3529-140.jpg)